streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -735,38 +753,51 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
     }
 
     @Override
+    @Deprecated
     public <VO, VR> KStream<K, VR> outerJoin(final KStream<K, VO> other,
                                              final ValueJoiner<? super V, ? super VO, ? extends VR> joiner,
                                              final JoinWindows windows,
                                              final Joined<K, V, VO> joined) {
-        return doJoin(other, joiner, windows, joined, new KStreamImplJoin(true, true));
+        final JoinedInternal<K, V, VO> joinedInternal = new JoinedInternal<>(joined);
+        final StreamJoined<K, V, VO> streamJoined =
+            StreamJoined.with(joinedInternal.keySerde(), joinedInternal.valueSerde(), joinedInternal.otherValueSerde()).withName(joinedInternal.name());
+        return outerJoin(other, joiner, windows, streamJoined);
+    }
+
+    @Override
+    public <VO, VR> KStream<K, VR> outerJoin(final KStream<K, VO> other,
+                                             final ValueJoiner<? super V, ? super VO, ? extends VR> joiner,
+                                             final JoinWindows windows,
+                                             final StreamJoined<K, V, VO> streamJoined) {
+
+        return doJoin(other, joiner, windows, streamJoined, new KStreamImplJoin(true, true));
     }
 
     private <VO, VR> KStream<K, VR> doJoin(final KStream<K, VO> other,
                                            final ValueJoiner<? super V, ? super VO, ? extends VR> joiner,
                                            final JoinWindows windows,
-                                           final Joined<K, V, VO> joined,
+                                           final StreamJoined<K, V, VO> streamJoined,
                                            final KStreamImplJoin join) {
         Objects.requireNonNull(other, "other KStream can't be null");
         Objects.requireNonNull(joiner, "joiner can't be null");
         Objects.requireNonNull(windows, "windows can't be null");
-        Objects.requireNonNull(joined, "joined can't be null");
+        Objects.requireNonNull(streamJoined, "streamJoined can't be null");
 
         KStreamImpl<K, V> joinThis = this;
         KStreamImpl<K, VO> joinOther = (KStreamImpl<K, VO>) other;
 
-        final JoinedInternal<K, V, VO> joinedInternal = new JoinedInternal<>(joined);
-        final NamedInternal name = new NamedInternal(joinedInternal.name());
+        final StreamJoinedInternal<K, V, VO> streamJoinedInternal = new StreamJoinedInternal<>(streamJoined);
+        final NamedInternal name = new NamedInternal(streamJoinedInternal.name());
```

Comment:
minor: 
```suggestion
        final NamedInternal name = new NamedInternal(streamJoinedInternal);
```

Should also work, it's a little less error-prone, since it results in NamedInternal extracting the name field defined by the Named interface, rather than us "manually" picking the right string field.

Reply:
Actually, this needs to stay as `NamedInternal` takes either a `Named` or `String` as the constructor parameter.  However I get your point here, but maybe we can defer this to a later follow-on PR.

Reply:
Hey, thanks for the reply... Am I missing something, isn't it true that a NamedInternal is also a Named, since `NamedInternal extends Named`?

Reply:
That's true,  but In this case, `StreamJoinedInternal` extends `StreamJoined` and `StreamJoined` implements `NamedOperation`.  
Now `Named` also implements `NamedOperation` but the `NamedInternal` constructor only accepts a `Named` or a `String`.  
So it seems we could do some follow up work to clean things up some.  But as you already said, I think that work is beyond the scope of this PR.

Reply:
Ah. I _was_ missing something. Sorry for the distraction; thanks for setting me straight.

Reply:
I had a similar though -- maybe we can make `NamedInternal` in interface with default method and let `StreamJoinedInternal implements NamedInternal` (and other classes). Also agree, this should be a follow up PR -- just make sure we don't forget about it :)